File: ocaml/lib/pickler_combinators.ml
type 'a pu = 'a -> unit
let pair pa pb = fun (a, b) -> pa a; pb b
let triple pa pb pc = fun (a, b, c) -> pa a; pb b; pc c
let quad pa pb pc pd = fun (a, b, c, d) -> pa a; pb b; pc c; pd d

let wrap (i, j) pa = fun x -> pa (j x); i x
let zero_to n = 
  let rec aux n = if n = 0 then (fun () -> 0) else wrap (fun x -> x/256, fun x -> x mod 256) (aux (n/256)) in
  aux n

let unit = fun () -> ()
let char = wrap (Char.code, Char.chr) (zero_to 255)
let bool = wrap (function true -> 1 | false -> 0, function 0 -> false | _ -> true) (zero_to 1)

let nat =
  let half = 128 in
  let rec aux n =
    if n < half then (fun x -> x)
    else wrap (fun x -> half + (x mod half), fun x -> x / half - 1) aux
  in
  aux

let fixed_list pa n =
  let rec aux n =
    if n = 0 then (fun () -> [])
    else wrap (fun (a, b) -> a :: b, fun l -> match l with a :: b -> (a, b) | [] -> failwith "impossible") (pair pa (aux (n-1)))
  in
  aux n

let list pa =
  let rec aux n =
    wrap (fun (n, l) -> l, fun l -> (List.length l, l)) (pair (nat n) (fixed_list pa n))
  in
  aux

let string = list char

let alt tag ps =
  let rec aux = function
    | [] -> failwith "alt: empty list"
    | p :: ps -> wrap (fun x -> (tag x, x), function (0, x) -> x | (n, x) -> aux ps (n-1, x)) p
  in
  aux (List.mapi (fun i p -> (i+1, p)) ps)

let p_maybe pa =
  alt (function None -> 0 | Some _ -> 1) [
    (fun () -> None);
    wrap (function x -> Some x, function None -> failwith "impossible" | Some x -> x) pa
  ]

let p_either pa pb =
  alt (function Left _ -> 0 | Right _ -> 1) [
    wrap (function x -> Left x, function Left x -> x | Right _ -> failwith "impossible") pa;
    wrap (function x -> Right x, function Left _ -> failwith "impossible" | Right x -> x) pb
  ]

File: haskell/lib/Pickler_combinators.hs
module Pickle (PU, pickle, unpickle, unit, char, bool, string, nat, zeroTo,
wrap, alt, pair, triple, quad, pMaybe, pEither, list) where
import CorePickle; import Maybe

pair :: PU a -> PU b -> PU (a,b)
pair pa pb = sequ fst pa (\ a ->
	sequ snd pb (\ b ->
	lift (a,b)))

triple :: PU a -> PU b -> PU c -> PU (a,b,c)
triple pa pb pc = sequ (\ (x,y,z) -> x) pa (\a ->
	sequ (\ (x,y,z) -> y) pb (\b ->
	sequ (\ (x,y,z) -> z) pc (\c ->
	lift (a,b,c))))

quad :: PU a -> PU b -> PU c -> PU d -> PU (a,b,c,d)
quad pa pb pc pd = sequ (\ (w,x,y,z) -> w) pa (\a ->
	sequ (\ (w,x,y,z) -> x) pb (\b ->
	sequ (\ (w,x,y,z) -> y) pc (\c ->
	sequ (\ (w,x,y,z) -> z) pd (\d ->
	lift (a,b,c,d)))))

wrap :: (a->b, b->a) -> PU a -> PU b
wrap (i,j) pa = sequ j pa (lift . i)

zeroTo :: Int -> PU Int
zeroTo 0 = lift 0
	zeroTo n = wrap (\ (hi,lo) -> hi * base + lo, (‘divMod‘ base))
	(pair (zeroTo (n ‘div‘ base)) belowBase)

unit :: PU ()
unit = lift ()

char :: PU Char
char = wrap (toEnum, fromEnum) (zeroTo 255)

bool :: PU Bool
bool = wrap (toEnum, fromEnum) (zeroTo 1)

nat :: PU Int
nat = sequ (\x -> if x < half then x else half + x ‘mod‘ half)

	belowBase
	(\lo -> if lo < half then lift lo
		else wrap (\hi->hi*half+lo, \n->n ‘div‘ half - 1) nat)
	where half = base ‘div‘ 2

fixedList :: PU a -> Int -> PU [a]
fixedList pa 0 = lift []
	fixedList pa n = wrap (\(a,b) -> a:b, \(a:b) -> (a,b)) (pair pa (fixedList pa (n-1)))

list :: PU a -> PU [a]
list = sequ length nat . fixedList

string :: PU String
string = list char

alt :: (a -> Int) -> [PU a] -> PU a
alt tag ps = sequ tag (zeroTo (length ps-1)) (ps !!)

pMaybe :: PU a -> PU (Maybe a)
pMaybe pa = alt tag [lift Nothing, wrap (Just, fromJust) pa]
	where tag Nothing = 0; tag (Just x) = 1

pEither :: PU a -> PU b -> PU (Either a b)
pEither pa pb = alt tag [wrap (Left, fromLeft) pa, wrap (Right, fromRight) pb]
	where tag (Left _) = 0; tag (Right _) = 1
		fromLeft (Left a) = a; fromRight (Right b) = b